elan/mux/mux.go

```diff
@@ -0,0 +1,110 @@
+// Package mux provides a multiplexer over two gocloud buckets.
+package mux
+
+import (
+	"context"
+	"io"
+
+	"github.com/peterebden/go-cli-init/v2"
+	"gocloud.dev/blob"
+)
+
+var log = cli.MustGetLogger()
+
+// New returns a new mux.
+func New(primary, secondary *blob.Bucket) *Mux {
+	return &Mux{
+		primary:   primary,
+		secondary: secondary,
+	}
+}
+
+type Mux struct {
+	primary, secondary *blob.Bucket
+}
+
+func (m *Mux) Exists(ctx context.Context, key string) (bool, error) {
+	if exists, err := m.primary.Exists(ctx, key); exists && err == nil {
+		return true, nil
+	}
+	return m.secondary.Exists(ctx, key)
+}
+
+func (m *Mux) NewRangeReader(ctx context.Context, key string, offset, length int64, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	if r, err := m.primary.NewRangeReader(ctx, key, offset, length, opts); err == nil {
+		return r, nil
+	}
+	// Try to read from the secondary. If successful, we will store the write in the primary
+	// on completion (but can only do that if we start at the beginning)
+	r, err := m.secondary.NewRangeReader(ctx, key, offset, length, opts)
+	if err != nil || offset != 0 {
+		return r, err
+	}
+	w, err := m.primary.NewWriter(ctx, key, nil)
+	if err != nil {
+		// This isn't fatal, we can still satisfy the read off the secondary.
+		log.Warning("Failed to duplicate blob to primary: %s", err)
+		return r, nil
+	}
+	return &reader{r: r, w: w}, nil
```

Comment:
Does this leave us open to partly written blobs? I guess that's not a problem but I thought i would check. Seems we're relying on the caller to subsequently create range-readers for all the byte-ranges of the blob in order to correctly write it back to the primary.

Reply:
I think it is OK because we always write to the end; above we check the offset and return there if it's not zero.
I don't think we have any use cases where someone writes a partial blob that isn't to the end (they might not start at the beginning if the upload was broken partway, in which case this duplication won't happen, but that shouldn't break anything).